hazelcast/src/hazelcast/client/Cluster.cpp

```diff
@@ -19,35 +19,45 @@
 
 
 #include "hazelcast/client/Cluster.h"
-#include "hazelcast/client/spi/ClusterService.h"
+#include "hazelcast/client/spi/ClientClusterService.h"
 #include "hazelcast/client/MembershipListener.h"
 #include "hazelcast/client/InitialMembershipListener.h"
 
 namespace hazelcast {
     namespace client {
-        Cluster::Cluster(spi::ClusterService &clusterService)
-        :clusterService(clusterService) {
+        Cluster::Cluster(spi::ClientClusterService &clusterService)
+                : clusterService(clusterService) {
+        }
 
+        void Cluster::addMembershipListener(MembershipListener *listener) {
+            clusterService.addMembershipListener(
+                    boost::shared_ptr<MembershipListener>(new MembershipListenerDelegator(listener)));
         }
 
-        void Cluster::addMembershipListener(InitialMembershipListener *listener) {
-            clusterService.addMembershipListener(listener);
+        bool Cluster::removeMembershipListener(MembershipListener *listener) {
+            return clusterService.removeMembershipListener(listener->getRegistrationId());
```

Comment:
For this to work, registration id needs to be set on listener. I could not where it is used.   
There is a test for this method `removeMembershipListener` but return value is not used. 
Can you add a return value check for that test ? And also make sure that `registration id ` is set to listener.